packages/rich-text/src/component/use-anchor.js

```diff
@@ -138,7 +140,12 @@ function getAnchor( editableContentElement, tagName, className ) {
  * @return {Element|VirtualAnchorElement|undefined|null} The active element or selection range.
  */
 export function useAnchor( { editableContentElement, settings = {} } ) {
```

Comment:
@ellatrix I would very much appreciate your opinion on this change.

The crux of this is that when clicking between multiple links in the same `contenteditable` (i.e. paragraph), the Link UI popover was being incorrectly anchored.

This was because the `useLayoutEffect` was not being re-run because all the props were the same between renders. This is because _all_ link formats have the same:

- `tagName`
- `className`
- `isActive` status

The last is is probably the one that will peek interest. Whilst it's obvious that the `tagName` will remain as `<a>` and the default for `className` is `""`, you'd think `isActive` would change when moving between links.

It does in fact do this, but when you click on a link the `isActive` status is set to `true`. If you then click _directly_ on another link the `isActive` status remains `true`. 

Ultimately I decided what we needed is a way to distinguish between mutliple link formats in a single contenteditable. It is for this reason that I have introduced the `activeAttributes` as an optional setting prop because this will be unique to each individual format. This guarantees the `useLayoutEffect` will run and recompute the anchor position correctly.

Reply:
@youknowriad also pointed out that https://github.com/WordPress/gutenberg/pull/58900/ seemed to make some large changes to the implementation of `useAnchor` which may be related to the incorrect anchor situation.

I tried restoring the previous implementation and there will still some edge cases around anchor positioning.

Reply:
I reverted `useAnchor` to the [state of the code](https://github.com/WordPress/gutenberg/blob/7e51b4352cdb5ad22b874d969f8dbb972d233525/packages/rich-text/src/component/use-anchor.js#L139C1-L180C2) prior to https://github.com/WordPress/gutenberg/pull/58900.

This appears to fix the anchor positioning issue without requiring any further changes to `useAnchor`.



https://github.com/WordPress/gutenberg/assets/444434/5208f872-3c65-4592-8f31-c20db138d69c

Therefore I wonder if @jeryj could confirm whether he notices any other regressions as a result of reverting https://github.com/WordPress/gutenberg/pull/58900?

It not then I propose we revert the changes in https://github.com/WordPress/gutenberg/pull/58900.

Reply:
Try creating a link :) 

https://github.com/WordPress/gutenberg/assets/967608/2057a931-ca9f-49f2-8404-4f6b1027da9b

Reply:
Ah yes it was about _creation_ of links. Right. Let's see if we can work this out then.

Reply:
Circling back, here's what I've done to resolve this:

- reverted (some of) the changes to the `useAnchor` API in https://github.com/WordPress/gutenberg/pull/58900 to restore the selection listeners present in WP 6.4.
- applied the internal changes around `isActive` from https://github.com/WordPress/gutenberg/pull/58900 which are a bug fix required to account for correct positioning when switching from "virtual" to Element anchors.
- reverted any changes to `useAnchor` API that I initially made when I introduced the `__unstableActiverAttributes` prop.

Rationale:

- changes to remove the listeners in https://github.com/WordPress/gutenberg/pull/58900 were made on the basis of them not being required for the Link UI implementation. However, as `useAnchor` is exported, we cannot know what use cases it is currently supporting and changing such a key implementation detail has the potential for backwards compatibility issues. Therefore reverting to the state of the code in WP 6.4 should not be considered a "change" for 6.5.
- changes to introduce the `isActive` prop were included in WP 6.5 prior to the Beta / RC period and thus are valid to be maintained. They are part of a legitimate bug fix.

From what I can see the functionality is now working as expected but I could do with careful checking.

Reply:
> we cannot know what use cases it is currently supporting and changing such a key implementation detail has the potential for backwards compatibility issues

It's not just about backward compatibility, having the listeners is the right fix to position the anchor properly when you switch between links using the mouse.

Reply:
Sounds like a reasonable way forwards to me. Thanks @getdave!